sdk/identity/azure-identity/azure/identity/aio/_internal/msal_credentials.py

```diff
@@ -0,0 +1,134 @@
+    from typing import Dict, Optional, Type, Union
+    from concurrent.futures import Executor
+
+from azure.core.credentials import AccessToken
+from azure.core.exceptions import ClientAuthenticationError
+import msal
+
+from .msal_transport_adapter import AsyncMsalTransportAdapter
+
+
+class MsalCredential(abc.ABC):
+    """Base class for async credentials wrapping MSAL applications"""
+
+    def __init__(
+        self, client_id: str, authority: str, client_credential: "Optional[Union[str, Dict[str,str]]]", **kwargs: "Any"
+    ) -> None:
+        self._authority = authority
+        self._client_credential = client_credential
+        self._client_id = client_id
+
+        self._lock = asyncio.Lock()
+        self._adapter = kwargs.pop("msal_adapter", None) or AsyncMsalTransportAdapter(**kwargs)
+
+        # postpone creating the wrapped application because its initializer uses the network
+        self._msal_app = None  # type: Optional[msal.ClientApplication]
+
+    @abc.abstractmethod
+    def get_token(self, *scopes: str) -> AccessToken:
+        pass
+
+    @abc.abstractmethod
+    def _get_app(self) -> msal.ClientApplication:
+        pass
+
+    async def _create_app(
+        self, cls: "Type[msal.ClientApplication]", executor: "Optional[Executor]" = None
+    ) -> msal.ClientApplication:
+        """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
+
+        if self._msal_app:
+            return self._msal_app
+
+        initializer = functools.partial(
+            cls,
+            client_id=self._client_id,
+            client_credential=self._client_credential,
+            authority=self._authority,
+            validate_authority=False,
+        )
+
+        loop = asyncio.get_event_loop()
+
+        # patch msal.authority so the MSAL app's init's network requests are received by the adapter
+        with mock.patch("msal.authority.requests", self._adapter) as adapter:
+            # We want the adapter to schedule transport on the event loop executing this method, so we need to pass
+            # the loop to the adapter. No argument passed to the MSAL app's init is passed through to the transport.
+            # So, we reluctantly use an instance attribute on the adapter. This should be safe because this method
+            # is called only from _get_app, and therefore within an asyncio Lock.
+            adapter.loop = loop
+            app = await loop.run_in_executor(executor, initializer)
+
+        # replace the client's requests.Session with adapter
+        app.client.session = self._adapter
+
+        return app
+
+
+class ConfidentialClientCredential(MsalCredential):
+    """Wraps an MSAL ConfidentialClientApplication with the TokenCredential API"""
+
+    async def get_token(self, *scopes: str, executor: "Optional[Executor]" = None):
+        """Get a token from the wrapped MSAL application"""
+
+        scopes = list(scopes)
```

Comment:
Why is a `tuple` not enough? Do you really need a list?

Reply:
MSAL raises if given anything else.

Reply:
...which is likely an example of someone going overboard with type checking... There may be a lesson in here somewhere :).

Reply:
A very minor improvement in this case might be to do the conversion only at the call site